go/fixchain/logger.go

```diff
@@ -182,7 +158,38 @@ func (l *Logger) postChain(p *toPost) {
 			} else {
 				log.Printf(ferr.Error.Error())
 				p.retries--
-				l.postToLog(p)
+				// postToLog() is called in a separate goroutine to avoid
+				// deadlock during retries.  Without the separate goroutine,
+				// deadlock can occur in the following situation:
+				//
+				// Suppose there is only one postServer() goroutine running, and
+				// it is blocked waiting for a toPost on the toPost chan.  A
+				// toPost gets added to the chan, which causes the following to
+				// happen:
+				// - the postServer takes the toPost from the chan.
+				// - the postServer calls l.postChain(toPost), and waits for
+				//   l.postChain() to return before going back to the toPost
+				//   chan for another toPost.
+				// - l.postChain() begins execution.  Suppose the first post
+				//   attempt of the toPost fails for some network-related
+				//   reason.
+				// - l.postChain retries and calls l.postToLog() to queue up the
+				//   toPost to try to post it again.
+				// - l.postToLog() tries to put the toPost on the toPost chan,
+				//   and blocks until a postServer takes it off the chan.
+				// But the one and only postServer is still waiting for
+				// l.postChain (and therefore l.postToLog) to return, and will
+				// not go to take another toPost off the toPost chan until that
+				// happens.
+				// Thus, deadlock.
+				//
+				// Similar situations with multiple postServers can easily be
+				// imagined.
+				l.wg.Add(1)
+				go func() {
+					l.postToLog(p)
+					l.wg.Done()
```

Comment:
1. Yes. Exactly.
2. Yes again, although I would say 'There's a race condition, where the Add(1) in postToLog happens after someone has successfully Wait()ed for that wg, which causes you to Add(1) before the go routine, and then requires a Done() inside the go routine to cancel the Add(1) before the goroutine.'  We need to add 1 to the wg as we are adding a new active request to the queue, and the Add(1) in postToLog() is what I consider to be that, as it's called right before we add the request to the queue.

But yes, if the repeated lines of code go in their own little async function, and we don't call postToLog in that function at all, then we can achieve all of these things in a much cleaner way!

Done, and it is much prettier!